be/src/storage/lake/vacuum.cpp

```diff
@@ -631,4 +633,210 @@ void delete_txn_log(TabletManager* tablet_mgr, const DeleteTxnLogRequest& reques
+}
+
+static StatusOr<std::map<std::string, DirEntry>> list_data_files(FileSystem* fs,
+                                                                 const std::string& segment_root_location,
+                                                                 int64_t expired_seconds) {
+    LOG(INFO) << "Start to list " << segment_root_location;
+    std::map<std::string, DirEntry> data_files;
+    int64_t total_files = 0;
+    int64_t total_bytes = 0;
+    const auto now = std::time(nullptr);
+    RETURN_IF_ERROR_WITH_WARN(ignore_not_found(fs->iterate_dir2(segment_root_location,
+                                                                [&](DirEntry entry) {
+                                                                    total_files++;
+                                                                    total_bytes += entry.size.value_or(0);
+
+                                                                    if (!is_segment(entry.name)) { // Only segment files
+                                                                        return true;
+                                                                    }
+                                                                    if (!entry.mtime.has_value()) {
+                                                                        LOG(WARNING) << "Fail to get modified time of "
+                                                                                     << entry.name;
+                                                                        return true;
+                                                                    }
+
+                                                                    if (now >= entry.mtime.value() + expired_seconds) {
+                                                                        data_files.emplace(entry.name, entry);
+                                                                    }
+                                                                    return true;
+                                                                })),
+                              "Failed to list " + segment_root_location);
+    LOG(INFO) << "Listed all data files, total files: " << total_files << ", total bytes: " << total_bytes
+              << ", candidate files: " << data_files.size();
+    return data_files;
+}
+
+static StatusOr<std::map<std::string, DirEntry>> find_orphan_data_files(FileSystem* fs, std::string_view root_location,
+                                                                        int64_t expired_seconds,
+                                                                        std::ostream& audit_ostream) {
+    const auto metadata_root_location = join_path(root_location, kMetadataDirectoryName);
+    const auto segment_root_location = join_path(root_location, kSegmentDirectoryName);
+
+    ASSIGN_OR_RETURN(auto metadatas, list_tablet_metadata(fs, metadata_root_location));
+
+    if (metadatas.empty()) {
```

Comment:
what if meta data are all cleaned but only data files remain, the data file may be not cleaned forever.

Reply:
fixed